### PR TITLE
Yd/signless with zero balance

### PR DESCRIPTION
**Pull request by YauheniDraichykau**

### CODE REVIEW

**Review by faizov — APPROVED** [review]

**Review by ereburg — APPROVED** [review]

File: frontend/packages/signless-transactions/src/hooks/use-create-session.ts [hunk 1]
```diff
@@ -76,16 +83,94 @@ function useCreateSession(programId: HexString, metadata: ProgramMetadata | unde
...
+      const { signature } = await signRaw({ address: account.address, data: hexToSign, type: 'bytes' });
+
+      const messageExtrinsic = getMessageExtrinsic({
+        CreateSession: { key: session.key, duration, allowedActions, signature },
+      });
+
+      const voucherExtrinsic = api.voucher.call(vId, { SendMessage: messageExtrinsic });
+
+      try {
+        voucherExtrinsic.signAndSend(pairToSave, statusCallback);
```
Comment (osipov-mit) [inline frontend/packages/signless-transactions/src/hooks/use-create-session.ts:166]:
```suggestion
        await voucherExtrinsic.signAndSend(pairToSave, statusCallback);
```

**Review by osipov-mit — CHANGES_REQUESTED** [review]

File: frontend/packages/ez-transactions/src/components/ez-transactions-switch/ez-transactions-switch.tsx [hunk 2]
```diff
@@ -21,7 +21,7 @@ function EzTransactionsSwitch() {
         shouldIssueVoucher={!gasless.isEnabled}
         disabled={!signless.isSessionActive && gasless.isActive}
         message={!signless.isSessionActive && gasless.isActive ? 'Gasless Session is Active' : ''}
-        requiredBalance={gasless.isEnabled ? 11 : undefined}
+        requiredBalance={gasless.isEnabled ? 0 : undefined}
```
[the remark this answers is not in the document]
Comment (YauheniDraichykau) [reply frontend/packages/ez-transactions/src/components/ez-transactions-switch/ez-transactions-switch.tsx:24]: there's a check on if we have undefined then we block signless functionality, but if we have 0 then it means the min limit to start signless is equal to 0, so user is able to enable it

File: frontend/packages/signless-transactions/src/hooks/use-create-session.ts [hunk 3]
```diff
@@ -76,16 +83,94 @@ function useCreateSession(programId: HexString, metadata: ProgramMetadata | unde
...
+    pair?: KeyringPair;
+  }>;
+  const handleStatus = (
+    { status, events }: ISubmittableResult,
+    { onSuccess = () => {}, onError = () => {}, onFinally = () => {} }: Options = {},
+  ) => {
+    if (!isApiReady) throw new Error('API is not initialized');
+
+    console.log('STATUS');
+    console.log(status.toHuman());
```
[the remark this answers is not in the document]
Comment (YauheniDraichykau) [reply frontend/packages/signless-transactions/src/hooks/use-create-session.ts]: Already removed this stuff

**Review by osipov-mit — APPROVED** [review]

File: frontend/packages/signless-transactions/src/hooks/use-create-session.ts [hunk 4]
```diff
@@ -76,16 +71,58 @@ function useCreateSession(programId: HexString, metadata: ProgramMetadata | unde
...
+
+      if (!signRaw) {
+        throw new Error('signRaw is not a function');
+      }
+
+      const key = decodeAddress(pairToSave.address);
+      const duration = session.duration;
+      const allowedActions = session.allowedActions;
+
+      const payloadToSign = { key, duration, allowedActions };
```
Comment (nikitayutanov) [inline frontend/packages/signless-transactions/src/hooks/use-create-session.ts:103]: We can just take `session` object?

File: frontend/packages/signless-transactions/src/hooks/use-create-session.ts [hunk 5]
```diff
@@ -76,16 +71,58 @@ function useCreateSession(programId: HexString, metadata: ProgramMetadata | unde
...
+
+      if (!metadata.types?.others?.output) {
```
Comment (nikitayutanov) [inline frontend/packages/signless-transactions/src/hooks/use-create-session.ts:105]: Would be better to throw an error there?

File: frontend/packages/signless-transactions/src/hooks/use-create-session.ts [hunk 6]
```diff
@@ -76,16 +71,58 @@ function useCreateSession(programId: HexString, metadata: ProgramMetadata | unde
...
+    onFinally: () => void;
+    pair?: KeyringPair;
+  }>;
 
   const createSession = async (
     session: Session,
     voucherValue: number,
-    { shouldIssueVoucher, ...options }: Options & { shouldIssueVoucher: boolean },
+    pairToSave: KeyringPair | null,
+    { shouldIssueVoucher, vId, ...options }: Options & { shouldIssueVoucher: boolean } & { vId: string | void },
```
Comment (nikitayutanov) [inline frontend/packages/signless-transactions/src/hooks/use-create-session.ts:85]:
```suggestion
    { shouldIssueVoucher, voucherId, ...options }: Options & { shouldIssueVoucher: boolean } & { voucherId: string | void },
```

File: frontend/packages/signless-transactions/src/hooks/use-create-session.ts [hunk 7]
```diff
@@ -76,16 +71,58 @@ function useCreateSession(programId: HexString, metadata: ProgramMetadata | unde
...
+
+        return;
+      }
+
+      const hexToSign = metadata.createType(metadata.types.others.output, payloadToSign).toHex();
+
+      const { signature } = await signRaw({ address: account.address, data: hexToSign, type: 'bytes' });
+
+      const messageExtrinsic = getMessageExtrinsic({
+        CreateSession: { key: session.key, duration, allowedActions, signature },
```
Comment (nikitayutanov) [inline frontend/packages/signless-transactions/src/hooks/use-create-session.ts:116]: If we can take `payloadToSign`  as a `session` object directly, why not to reuse it's there also?

File: frontend/packages/signless-transactions/src/hooks/use-create-session.ts [hunk 8]
```diff
@@ -76,16 +71,58 @@ function useCreateSession(programId: HexString, metadata: ProgramMetadata | unde
...
+      });
+
+      const voucherExtrinsic = api.voucher.call(vId, { SendMessage: messageExtrinsic });
+
+      await sendTransaction(voucherExtrinsic, pairToSave, ['UserMessageSent'], { ...options, onError });
+
+      return;
```
Comment (nikitayutanov) [inline frontend/packages/signless-transactions/src/hooks/use-create-session.ts:123]:
```suggestion
      return sendTransaction(voucherExtrinsic, pairToSave, ['UserMessageSent'], { ...options, onError });
```

File: frontend/packages/signless-transactions/src/hooks/use-create-session.ts [hunk 9]
```diff
@@ -100,10 +137,13 @@ function useCreateSession(programId: HexString, metadata: ProgramMetadata | unde
...
 
-    const messageExtrinsic = getMessageExtrinsic({ DeleteSessionFromAccount: null });
+    const messageExtrinsic = getMessageExtrinsic({
+      DeleteSessionFromAccount: null,
+    });
+
     const txs = [messageExtrinsic];
 
-    const voucher = await getLatestVoucher(key);
+    const voucher = await getLatestVoucher(key); //if account has voucher
```
Comment (nikitayutanov) [inline frontend/packages/signless-transactions/src/hooks/use-create-session.ts:146]:
```suggestion
    const voucher = await getLatestVoucher(key);
```

I think no need it this comment, code pretty much is descriptive enough.

File: frontend/packages/signless-transactions/src/utils.ts [hunk 10]
```diff
@@ -4,6 +4,13 @@ import { SubmittableExtrinsic } from '@polkadot/api/types';
 import { encodeAddress } from '@polkadot/keyring';
 import { KeyringPair$Json, KeyringPair } from '@polkadot/keyring/types';
 
+type Options = Partial<{
+  onSuccess: () => void;
+  onError: (error: string) => void;
+  onFinally: () => void;
+  pair?: KeyringPair;
+}>;
```
Comment (nikitayutanov) [inline frontend/packages/signless-transactions/src/utils.ts:12]: Same as above.

**Review by nikitayutanov — CHANGES_REQUESTED** [review]

**Comment by nikitayutanov** [comment]
Also, I think it's better to avoid `console.log`s alltogether. At this point I believe we just need to find some normal logging solution.

File: frontend/packages/signless-transactions/src/hooks/use-create-session.ts (hunk 8 above)
Comment (YauheniDraichykau) [reply frontend/packages/signless-transactions/src/hooks/use-create-session.ts:123]: what's the point of returning sendTransaction ?

Comment (nikitayutanov) [reply frontend/packages/signless-transactions/src/hooks/use-create-session.ts:123]: Same behavior as to explicitly await it first and make return later.

**Review by nikitayutanov — APPROVED** [review]

**Pull request merged**